pgmpy/models/DynamicBayesianNetwork.py

```diff
@@ -0,0 +1,500 @@
+
+import numpy as np
+from pgmpy.models import BayesianModel
+from pgmpy.factors import TabularCPD, TreeCPD, RuleCPD
+from collections import OrderedDict, namedtuple
+
+
+edge_type = namedtuple('edge_type', ('start', 'end'))
+
+
+class DynamicBayesianNetwork(BayesianModel):
+    def __init__(self, ebunch=None):
+        """
+        Base class for Dynamic Bayesian Network
+
+        This model is a time variant of the static Bayesian model, where each timeslice has
+        some static nodes and is then replicated over a certain time slice.
+        
+        The nodes can be hashable python objects.
+
+        However, the hidden nodes will compulsory have the following form.
+        (node_name, time_slice)
+
+        Here, node_name is the node that is inserted
+        while the time_slice is an integer value, which denotes
+        the index of the time_slice that the node belongs to.
+
+        Edges are represented as links between the nodes.
+
+        Parameters:
+        ----------
+        data: Data to initialize graph.  If data=None (default) an empty
+              graph is created.  The data can be an edge list, or any
+              NetworkX graph object
+
+        Examples:
+        --------
+        Create an empty Dynamic Bayesian Network with no nodes and no edges
+        >>> from pgmpy.models import DynamicBayesianNetwork as DBN
+        >>> dbn = DBN()
+        adding nodes and edges inside the dynamic bayesian network. A single
+        node can be added using the method below.
+        >>> dbn.add_nodes_from(['D','G','I','S','L'])
+        >>> dbn.add_edges_from([(('D',0),('G',0)),(('I',0),('G',0)),(('G',0),('L',0))])
+
+        most of the methods will be imported from Bayesian Model
+        However there will be a few changes
+        >>> dbn.list_of_nodes
+        ['D','G','I','S','L']
+        >>> dbn.edges()
+        [(('D',0),('G',0)),(('I',0),('G',0)),(('G',0),('L',0))]
+        If some edges connect nodes not yet in the model, the nodes
+        are added automatically.  There are no errors when adding
+        nodes or edges that already exist.
+
+        Methods:
+        -------
+        add_nodes_from
+        add_edge
+        add_edges_from
+        intra_slice
+        inter_slice
+        add_cpds
+        get_cpds
+        compute_initial_state
+        unroll
+        remove_timeslice
+        """
+        super().__init__()
+        self.inter_edges = []
+        self.intra_edges = []
+        self.dbn_cpds = OrderedDict()
+        self.list_of_nodes = []
+        if ebunch:
+            self.add_nodes_from(ebunch)
+        self.timestate = 1
+        self.computed = False
+
+    def add_nodes_from(self, nodes):
```

Comment:
There is a major issue with ```add_nodes_from``` function. We have not defined a ```add_node``` method in this class, so when someone tries to use call ```add_node()``` method it would be using the ```add_node``` method from ```BayesianModel``` class. For example 

```python
G = DynamicBayesianModel()
G.add_node('A')
```

it won't be adding ```A``` in ```list_of_nodes```, as there is no such variable in ```BayesianModel```. It would lead to inconsistency. So my suggestion is don't define ```add_node``` and ```add_nodes_from``` until you want to have some extra constraints on the nodes. And whenever you want to access the nodes instead of using ```list_of_nodes``` just call ```self.nodes()```.